caddy.go

```diff
@@ -186,9 +187,26 @@ func (i *Instance) Restart(newCaddyfile Input) (*Instance, error) {
 	i.wg.Add(1)
 	defer i.wg.Done()
 
+	var err error
+	// if something went wrong on restart then run onRestartFailed callbacks
+	defer func() {
+		r := recover()
+		if err != nil || r != nil {
+			for _, fn := range i.onRestartFailed {
+				err = fn()
+				if err != nil {
+					return
+				}
```

Comment:
I see several issues here.
1. The actual error from `OnRestartFailed` handler is ignored, i.e. you use return value as boolean type. Consider at least logging the error.
2. If one of `OnRestartFailed` handlers fails (returns error) other handlers might not be called because you exit the loop. This means that you don't achieve your goal as described in #2250 - 
> So every OnRestart for plugin would finish by either OnShutdow() or OnRestartFailed().

 **Note**, that with current `onShutdown` related code
https://github.com/mholt/caddy/blob/d3e3fc533f7046d47d60818234f3bb50d62ea1b6/caddy.go#L233-L238
by the similar reasons both `OnShutdown()` and `OnRestartFailed()` handlers can be called for the same plugin. This could happen if `OnShutdown()` succeded for plugin _A_ but failed for plugin _B_, i.e. bot `OnShutdown()` and `OnRestartFailed()` handlers of plugin _A_ will be called
3. If panic is caught and one of `onRestartFailed` handlers fails (returns error) you won't re-issue panic at line 202, because you have exited at line 198

Reply:
1: I don't see why it should be ignored. This is defered function and err is assigned to a new value.
2, 3: Agree with your statements, but I am wondering what should be the most correct behavior here? Should I compose errors somehow? Or just log them right on place? @mholt, @elcore, any preferences?

Reply:
`defer` assigns new value to variable `err` too late. Statement `return` copies the original `err` value to return value before calling deferred functions. You can use named return value if you want deferred function to change the return value. See example at https://play.golang.org/p/QCJBUPc7ePK

Reply:
You are right. I don't use named value anymore here. So, return should be changed.

Reply:
Made the log message. Usef format like in https://github.com/mholt/caddy/blob/master/sigtrap.go#L83 .